drv/grapefruit-seq-server/src/main.rs

```diff
@@ -0,0 +1,266 @@
+        //
+        // SP_TO_FPGA_CFG_CLK / SP_TO_FPGA_CFG_DAT
+        // This is on SPI2, port B
+        //
+        // Wait for DONE (FPGA_TO_SP_CONFIG_DONE)
+
+        // Bind to the sequencer device on our SPI port
+        let seq = spi.device(drv_spi_api::devices::FPGA);
+
+        // TODO do we need to send the bus width / synchronization word / device
+        // ID ourselves, or are they built into the image?
+
+        let blob = aux
+            .get_blob_by_tag(*b"FPGA")
+            .map_err(|_| SeqError::AuxMissingBlob)?;
+        let mut scratch_buf = [0u8; 128];
+        let mut pos = blob.start;
+        let mut sha = Sha3_256::new();
+        let mut decompressor = gnarle::Decompressor::default();
+        while pos < blob.end {
+            let amount = (blob.end - pos).min(scratch_buf.len() as u32);
+            let chunk = &mut scratch_buf[0..(amount as usize)];
+            if let Err(e) = aux.read_slot_with_offset(blob.slot, pos, chunk) {
+                return Err(SeqError::AuxReadError(e));
+            }
+            sha.update(&chunk);
+            pos += amount;
+
+            // Reborrow as an immutable chunk, then decompress
+            let mut chunk = &scratch_buf[0..(amount as usize)];
+            let mut decompress_buffer = [0; 512];
+
+            while !chunk.is_empty() {
+                let decompressed_chunk = gnarle::decompress(
+                    &mut decompressor,
+                    &mut chunk,
+                    &mut decompress_buffer,
+                );
+
+                // The compressor may have encountered a partial run at the
+                // end of the `chunk`, in which case `decompressed_chunk`
+                // will be empty since more data is needed before output is
+                // generated.
+                if !decompressed_chunk.is_empty() {
+                    // Write the decompressed bitstream to the FPGA over SPI
+                    seq.write(&decompressed_chunk)
+                        .map_err(|e| SeqError::SpiWrite(e))?;
+                    ringbuf_entry!(Trace::ContinueBitstreamLoad(
+                        decompressed_chunk.len()
+                    ));
+                }
+            }
+        }
+
+        let sha_out: [u8; 32] = sha.finalize().into();
+        if sha_out != gen::FPGA_BITSTREAM_CHECKSUM {
+            // Reset the FPGA to clear the invalid bitstream
+            sys.gpio_reset(FPGA_PROGRAM_L);
+            hl::sleep_for(1);
+            sys.gpio_set(FPGA_PROGRAM_L);
+
+            return Err(SeqError::AuxChecksumMismatch);
+        }
+
+        // Wait for the FPGA to pull DONE high
+        //
+        // This pin is by default an input, and we never change it
+        loop {
+            let done = sys.gpio_read(FPGA_CONFIG_DONE) != 0;
+            if done {
+                break;
+            } else {
+                ringbuf_entry!(Trace::WaitForDone);
+                hl::sleep_for(2);
+            }
+        }
```

Comment:
nit, not a big deal: isn't this a `while` loop?
```suggestion
        while sys.gpio_read(FPGA_CONFIG_DONE) != 0 {
            ringbuf_entry!(Trace::WaitForDone);
            hl::sleep_for(2);
        }
```